File: typescript/src/crypto/hmac.ts
/**
 * HMAC-SHA-256 signature backend.
 *
 * Profile:
 *   - Algorithm: HMAC-SHA-256 (RFC 2104).
 *   - Secret length: **minimum 32 bytes** enforced at import / generate.
 *     Shorter secrets are rejected with `InsufficientKeyStrength`.
 *   - Signature size: 32 bytes.
 *   - Verification uses `crypto.timingSafeEqual` — never `===` — to avoid
 *     timing side-channels.
 *
 * Caveats (documented for operators):
 *   - HMAC is **symmetric**. The "public key" is the same secret as the
 *     "private key". Distributing the secret to multiple verifiers means
 *     each verifier can also forge tokens. Use only in self-contained
 *     deployments where issuer and verifier are the same trust boundary
 *     (e.g., a single process, or a cluster sharing one signing key over
 *     trusted transport).
 *   - PEM is not a natural representation for HMAC; this backend accepts
 *     raw bytes only. The `PemKeyMaterial` input path is rejected with a
 *     clear error so operators can't accidentally wire a PEM key into HMAC.
 */

import { createHmac, randomBytes, timingSafeEqual } from 'node:crypto';

import { errors } from '../errors.ts';
import type {
  KeyGenerator,
  PemKeyMaterial,
  PrivateKeyHandle,
  PublicKeyHandle,
  RawKeyMaterial,
  SignatureBackend,
  Signer,
  Verifier,
} from './types.ts';

export const HMAC_MIN_SECRET_LEN = 32;
export const HMAC_SIG_LEN = 32;
export const HMAC_DEFAULT_SECRET_LEN = 32;
export const HMAC_HASH = 'sha256' as const;

interface HMACHandle extends PrivateKeyHandle, PublicKeyHandle {
  readonly __hmacSecret: Uint8Array;
}

function wrap(secret: Uint8Array): HMACHandle {
  return { __hmacSecret: secret } as HMACHandle;
}
function unwrap(h: PrivateKeyHandle | PublicKeyHandle): Uint8Array {
  return (h as HMACHandle).__hmacSecret;
}

function isRaw(m: PemKeyMaterial | RawKeyMaterial): m is RawKeyMaterial {
  return 'publicRaw' in m && m.publicRaw instanceof Uint8Array;
}

function assertSecretStrength(secret: Uint8Array): void {
  if (secret.length < HMAC_MIN_SECRET_LEN) {
    throw errors.insufficientKeyStrength(
      `HMAC secret must be ≥ ${HMAC_MIN_SECRET_LEN} bytes, got ${secret.length}`,
    );
  }
}

class HMACBackend implements SignatureBackend, Signer, Verifier, KeyGenerator {
  readonly alg = 'hs256' as const;

  async importPrivate(
    material: PemKeyMaterial | RawKeyMaterial,
    _passphrase?: string,
  ): Promise<PrivateKeyHandle> {
    if (!isRaw(material)) {
      throw errors.tokenMalformed('HMAC does not support PEM material — pass raw bytes');
    }
    const secret = material.privateRaw ?? material.publicRaw;
    assertSecretStrength(secret);
    // Copy so later mutation of the caller's array can't invalidate the handle.
    return wrap(new Uint8Array(secret));
  }

  async importPublic(material: PemKeyMaterial | RawKeyMaterial): Promise<PublicKeyHandle> {
    if (!isRaw(material)) {
      throw errors.tokenMalformed('HMAC does not support PEM material — pass raw bytes');
    }
    // For a symmetric scheme the "public" handle is the same secret. This
    // intentionally mirrors importPrivate so the signature backend interface
    // stays uniform across algorithms.
    assertSecretStrength(material.publicRaw);
    return wrap(new Uint8Array(material.publicRaw));
  }

  async sign(key: PrivateKeyHandle, data: Uint8Array): Promise<Uint8Array> {
    const mac = createHmac(HMAC_HASH, unwrap(key)).update(data).digest();
    return new Uint8Array(mac.buffer, mac.byteOffset, mac.byteLength);
  }

  async verify(key: PublicKeyHandle, data: Uint8Array, signature: Uint8Array): Promise<boolean> {
    const expected = createHmac(HMAC_HASH, unwrap(key)).update(data).digest();
    if (signature.length !== expected.length) return false;
    // timingSafeEqual requires both to be Buffers of identical length.
    return timingSafeEqual(Buffer.from(signature), expected);
  }

  async generate(_passphrase: string): Promise<{ pem: PemKeyMaterial; raw: RawKeyMaterial }> {
    const secret = new Uint8Array(randomBytes(HMAC_DEFAULT_SECRET_LEN));
    // No PEM representation for HMAC; we surface null/empty strings and
    // require downstream to consume `raw`. The core's `KeyRecord` type
    // allows `privatePem: null`, and `publicPem` is documented as "same
    // bytes as privateRaw" for the symmetric case.
    return {
      pem: {
        privatePem: null,
        publicPem: '',
      },
      raw: {
        privateRaw: secret,
        publicRaw: secret,
      },
    };
  }
}

export const hmacBackend: SignatureBackend = new HMACBackend();

File: licensing/crypto/hmac/hmac.go
// Package hmac implements the licensing SignatureBackend for HMAC-SHA-256.
//
// Uses only the Go standard library (crypto/hmac, crypto/sha256).
// Mirrors @licensing/core/crypto/hmac semantically:
//
//   - Secrets must be at least 32 bytes; shorter secrets are rejected with
//     ErrInsufficientKeyStrength
//   - MAC is 32 bytes (SHA-256 output)
//   - Verification uses hmac.Equal (constant-time) — never bytes.Equal
//
// Caveats (documented for operators — same text as the TS sibling):
//
//   - HMAC is SYMMETRIC. The "public" key is the same secret as the
//     "private" key. Distributing the secret to multiple verifiers means
//     each verifier can also forge tokens. Use only in self-contained
//     deployments where issuer and verifier are the same trust boundary.
//   - PEM is not a natural representation for HMAC; this backend accepts
//     raw bytes only. The PEM input path is rejected with a clear error
//     so operators can't accidentally wire a PEM key into HMAC.
package hmac

import (
	stdhmac "crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"fmt"

	lic "github.com/AnoRebel/licensing/licensing"
)

// Profile parameters — must match @licensing/core/crypto/hmac.ts.
const (
	MinSecretLen     = 32
	DefaultSecretLen = 32
	SignatureLen     = 32 // = SHA-256 output
)

// Backend is the sole SignatureBackend for HMAC-SHA-256. Stateless.
type Backend struct{}

// New returns a ready-to-register backend value.
func New() Backend { return Backend{} }

// Alg returns the algorithm identifier this backend handles.
func (Backend) Alg() lic.KeyAlg { return lic.AlgHS256 }

// handle holds the shared secret. Since HMAC is symmetric, both
// PrivateKeyHandle and PublicKeyHandle are backed by the same struct.
type handle struct {
	secret []byte
}

// ImportPrivate loads a secret from raw bytes. PEM input is explicitly
// rejected — there is no natural PEM representation for an HMAC secret.
func (Backend) ImportPrivate(m lic.KeyMaterial, _ string) (lic.PrivateKeyHandle, error) {
	if m.Pem.PrivatePem != "" || m.Pem.PublicPem != "" {
		return nil, wrapMalformed("hmac does not support PEM material — pass raw bytes")
	}
	// Prefer PrivateRaw; fall back to PublicRaw since they're the same secret.
	secret := m.Raw.PrivateRaw
	if len(secret) == 0 {
		secret = m.Raw.PublicRaw
	}
	if len(secret) == 0 {
		return nil, wrapMalformed("hmac: raw secret not provided")
	}
	if len(secret) < MinSecretLen {
		return nil, wrapStrength(fmt.Sprintf(
			"HMAC secret must be ≥ %d bytes, got %d", MinSecretLen, len(secret)))
	}
	cp := make([]byte, len(secret))
	copy(cp, secret)
	return &handle{secret: cp}, nil
}

// ImportPublic mirrors ImportPrivate for the symmetric case.
func (Backend) ImportPublic(m lic.KeyMaterial) (lic.PublicKeyHandle, error) {
	if m.Pem.PrivatePem != "" || m.Pem.PublicPem != "" {
		return nil, wrapMalformed("hmac does not support PEM material — pass raw bytes")
	}
	secret := m.Raw.PublicRaw
	if len(secret) == 0 {
		secret = m.Raw.PrivateRaw
	}
	if len(secret) == 0 {
		return nil, wrapMalformed("hmac: raw secret not provided")
	}
	if len(secret) < MinSecretLen {
		return nil, wrapStrength(fmt.Sprintf(
			"HMAC secret must be ≥ %d bytes, got %d", MinSecretLen, len(secret)))
	}
	cp := make([]byte, len(secret))
	copy(cp, secret)
	return &handle{secret: cp}, nil
}

// Sign returns the 32-byte HMAC-SHA-256 of data under the handle's secret.
func (Backend) Sign(h lic.PrivateKeyHandle, data []byte) ([]byte, error) {
	hh, ok := h.(*handle)
	if !ok {
		return nil, wrapMalformed(fmt.Sprintf(
			"hmac.Sign: wrong handle type %T", h))
	}
	m := stdhmac.New(sha256.New, hh.secret)
	m.Write(data)
	return m.Sum(nil), nil
}

// Verify returns (true, nil) iff sig matches the expected HMAC of data under
// the handle's secret. A structurally wrong-length sig returns (false, nil)
// without touching the secret material. Comparison is constant-time.
func (Backend) Verify(h lic.PublicKeyHandle, data, sig []byte) (bool, error) {
	hh, ok := h.(*handle)
	if !ok {
		return false, wrapMalformed(fmt.Sprintf(
			"hmac.Verify: wrong handle type %T", h))
	}
	if len(sig) != SignatureLen {
		return false, nil
	}
	m := stdhmac.New(sha256.New, hh.secret)
	m.Write(data)
	expected := m.Sum(nil)
	return stdhmac.Equal(expected, sig), nil
}

// Generate produces a fresh random 32-byte secret. The returned PemKeyMaterial
// is empty because HMAC has no natural PEM form; callers consume the raw
// bytes.
func (Backend) Generate(_ string) (lic.PemKeyMaterial, lic.RawKeyMaterial, error) {
	secret := make([]byte, DefaultSecretLen)
	if _, err := rand.Read(secret); err != nil {
		return lic.PemKeyMaterial{}, lic.RawKeyMaterial{}, err
	}
	// Both halves point at copies of the same secret so consumers who reach
	// for PrivateRaw or PublicRaw both work, and mutating one doesn't corrupt
	// the other.
	privCopy := make([]byte, DefaultSecretLen)
	pubCopy := make([]byte, DefaultSecretLen)
	copy(privCopy, secret)
	copy(pubCopy, secret)
	return lic.PemKeyMaterial{},
		lic.RawKeyMaterial{PrivateRaw: privCopy, PublicRaw: pubCopy},
		nil
}

// -----------------------------------------------------------------------
// Raw-bytes accessor.
// -----------------------------------------------------------------------

// Secret returns a copy of the raw secret bytes held by a handle.
func Secret(h any) ([]byte, error) {
	hh, ok := h.(*handle)
	if !ok {
		return nil, wrapMalformed(fmt.Sprintf(
			"hmac.Secret: wrong handle type %T", h))
	}
	out := make([]byte, len(hh.secret))
	copy(out, hh.secret)
	return out, nil
}

// Compile-time assertion.
var _ lic.SignatureBackend = Backend{}

// -----------------------------------------------------------------------
// Error wrappers — same pattern as the ed25519 / rsa backends.
// -----------------------------------------------------------------------

func wrapStrength(msg string) error {
	return fmt.Errorf("%w: %s", lic.ErrInsufficientKeyStrength, msg)
}

func wrapMalformed(msg string) error {
	return fmt.Errorf("%w: %s", lic.ErrTokenMalformed, msg)
}
